src/blockchain/compact_filters/p2p.rs

```diff
@@ -0,0 +1,646 @@
+        let other_node_count = self.directory.get_non_cbf_address_count();
+
+        DiscoveryData {
+            queued: queued_count,
+            visited: visited_count,
+            non_cbf_count: cbf_node_count + other_node_count,
+            cbf_count: cbf_node_count,
+        }
+    }
+
+    /// Spawn [self.thread] no. of worker threads
+    fn spawn_workers(&mut self) -> Vec<JoinHandle<()>> {
+        let mut worker_handles: Vec<JoinHandle<()>> = vec![];
+        for _ in 0..self.threads {
+            let sender = self.sender.clone();
+            let discovery = self.discovery.clone();
+            let network = self.network;
+            let worker_handle = thread::spawn(move || {
+                let mut worker = AddressWorker::new(discovery, sender, network);
+                worker.work();
+            });
+            worker_handles.push(worker_handle);
+        }
+        worker_handles
+    }
+
+    /// Crawl the Bitcoin network until given number of cbf/non_cbf nodes are found
+    ///
+    /// - This will start a bunch of crawlers.
+    /// - load up the existing cache.
+    /// - Update the cache with new found peers.
+    /// - check if address is in banlist
+    /// - run crawlers until buffer requirement is matched
+    /// - flush the current cache into disk
+    pub fn fetch(&mut self) {
+        self.spawn_workers();
+
+        // Get already existing cache
+        let mut cache = match AddressCache::from_file(&self.cache_filename) {
+            Some(cache) => cache,
+            None => AddressCache::empty(),
+        };
+
+        while self.directory.get_cbf_buffer() < self.cbf_buffer
+            || self.directory.get_non_cbf_buffer() < self.non_cbf_buffer
+        {
+            if let Ok(message) = self.receiver.recv() {
+                match message {
+                    AddressMessage::NewValid(addr, flag) => {
+                        if !cache.banned_peers.contains(&addr) {
+                            let cbf = flag.has(ServiceFlags::COMPACT_FILTERS);
+
+                            self.directory.add_address(addr, cbf);
+
+                            cache.add_address(addr, cbf);
+                        }
+                    }
+                }
+            }
+            self.progress.update(self.get_progress()).unwrap();
+        }
+
+        // When completed, flush the cache
+        cache.write_to_file(&self.cache_filename).unwrap();
+    }
+
+    /// Get a new addresses not previously provided
+    pub fn get_new_cbf_address(&mut self) -> SocketAddr {
+        self.directory.get_new_address(true)
+    }
+
+    /// Get a new non_cbf address
+    pub fn get_new_non_cbf_address(&mut self) -> SocketAddr {
+        self.directory.get_new_address(false)
+    }
+
+    /// Ban an address
+    pub fn ban_peer(&mut self, addrs: &SocketAddr, cbf: bool) {
+        let mut cache = AddressCache::from_file(&self.cache_filename).unwrap();
+
+        cache.add_to_banlist(*addrs, cbf);
```

Comment:
I think this is not being written back to disk. Adding the cache as a struct member would allow to reuse it here as well

Reply:
Ah write, it wasn't being written to disk. Fixed that. 

Because of the reasons explained here https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660353392 I am inclining against adding the cache to the structure.